### PR TITLE
Fix #9: Choose project based on active file

### LINKED ISSUE #9: Filter the list of projects in Mechanic.fs to the ones that contain the currently opended file
/cc @Krzysztof-Cieslak

### PR DESCRIPTION
**Pull request by MangelMaxime**

### CODE REVIEW

**Comment by MangelMaxime** [comment]
Careful, don't release this until we fixed #15 or we need a manual action on release.

File: src/Mechanic.fs [hunk 1]
```diff
@@ -50,16 +73,12 @@ let activate (context : Vscode.ExtensionContext) =
...
-                |> Seq.toList
-                |> List.exists(fun currentProject ->
-                    currentProject.Project = project.Project
-                )
-
-            // Only add the project if not already known
-            if not exist then
-                state.Add project
+            // If the project is already known, update it
+            if state.ContainsKey project.Project then
```
Comment (Krzysztof-Cieslak) [inline src/Mechanic.fs]: I think there is no need for this check. ` state.[project.Project] <- project` should just add item if it doesn't exists, as far as I remeber

Comment (MangelMaxime) [reply src/Mechanic.fs]: Let me try

**Comment by MangelMaxime** [comment]
Don't merge this PR yet, I should filter current file type.

- If the file end with `.fsproj` use it directly
- If the file end with `.fs` try to find the associated project
- For others, will go for the default way

File: src/Mechanic.fs (hunk 1 above)
Comment (MangelMaxime) [reply src/Mechanic.fs]: Indeed this seems to work

**Comment by forki** [comment]
@Krzysztof-Cieslak when this is ready - please release ;-)
I'm away from PC now

**Comment by MangelMaxime** [comment]
Should be ready for review.

@forki Perhaps, you could give us the action needed to release a new version to avoid deploying a broken version of mech.dll

**Comment by forki** [comment]
Ah good point.
Easiest fix. Change the build script here to run dotnet publish in the src
folder of mechanic.commandline. Just right before that line that copies the
mech libs. And instead of copying them from debug folder, copy them from
the /public folder within that debug folder.
Or even better: do the above but for release mode.

Maxime Mangel <notifications@github.com> schrieb am Di., 13. März 2018,
17:50:

**Comment by MangelMaxime** [comment]
I sent a PR https://github.com/fsprojects/Mechanic/pull/86

**Comment by forki** [comment]
Even better

**Comment by forki** [comment]
is this ready?

**Comment by MangelMaxime** [comment]
Yes, it's ready for review and test.

**Comment by forki** [comment]
I think we broke something

tried on mechanic itself:

image

**Comment by MangelMaxime** [comment]
Was the project loaded ?
Does the file belongs to an fsproj ?

**Comment by MangelMaxime** [comment]
Seems like we were in this case:

https://github.com/ionide/ionide-vscode-mechanic/blob/98c8d1ad5c9b8548df1dfcc18321e0399251ccab/src/Mechanic.fs#L49-L54

In this case we probably need to execute:

https://github.com/ionide/ionide-vscode-mechanic/blob/98c8d1ad5c9b8548df1dfcc18321e0399251ccab/src/Mechanic.fs#L64-L85

**Comment by forki** [comment]
PR appreciated ;-)

**Comment by MangelMaxime** [comment]
Hum...

If the file is not associated to project file then running Mechanic is useless. We probably need to add the file to an fsproj first.

What do you think ?

**Comment by forki** [comment]
it is in the project. 

The filter                     |> List.contains textEditor.document.uri.path dosesn't work.

I'm debugging it

**Comment by forki** [comment]
Ok I listed the files in the project and then the textEditor.document.uri.path:

image

obviously they are in different format. similar issue to https://github.com/ionide/ionide-vscode-mechanic/issues/18

do we already have a way to normilze that? /cc @Krzysztof-Cieslak

**Comment by forki** [comment]
how do we know if we are on windows or linux?

**Comment by MangelMaxime** [comment]
@forki Ah yes logic...

We can normalize the path: https://nodejs.org/api/path.html#path_path_normalize_path
We can detect the OS: 

```fs
open Fable.Import

if Node.Exports.os.platform() = Node.Base.NodeJS.Win32 then
    // We are under windows
```

**Comment by Krzysztof-Cieslak** [comment]
I think `document.filename` or `document.uri.fsPath` should have correct, normalized paths (from top of my head)

**Comment by MangelMaxime** [comment]
@forki Can I let you check that ? I don't have access to a Windows computer

**Comment by forki** [comment]
current solution works for me. maybe we should just keep it

**Pull request merged**